pkg/controller/epgcache.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2021 Cisco Systems, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"github.com/noironetworks/aci-containers/pkg/apicapi"
+	"github.com/noironetworks/aci-containers/pkg/metadata"
+
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+func (cont *AciController) clearFaultInstances() {
+
+	queryfilter := fmt.Sprintf("query-target-filter=and(wcard(vmmClusterFaultInfo.dn,\"comp/prov-%s/ctrlr-\\[%s\\]-%s\"))",
+		cont.vmmDomainProvider(), cont.config.AciVmmDomain, cont.config.AciVmmDomain)
+	args := []string{
+		queryfilter,
+		"order-by=vmmClusterFaultInfo.modTs|desc",
+	}
+
+	faultUri := fmt.Sprintf("/api/node/class/vmmClusterFaultInfo.json?%s", strings.Join(args, "&"))
+	apicresp, err := cont.apicConn.GetApicResponse(faultUri)
+	if err != nil {
+		return
+	}
+	for _, obj := range apicresp.Imdata {
+		for _, body := range obj {
+			dn, ok := body.Attributes["dn"].(string)
+			if ok {
+				cont.indexMutex.Lock()
+				cont.apicConn.DeleteDn(dn)
+				cont.log.Debug("Clearing existing Fault Instance: ", dn)
+				cont.indexMutex.Unlock()
+			}
+		}
+	}
+	return
+}
+
+func (cont *AciController) getSubString(dn string) string {
+	delimiterString := fmt.Sprintf("uni/vmmp-%s/dom-%s/eppd-[", cont.vmmDomainProvider(), cont.config.AciVmmDomain)
+	epgDn := strings.Split(dn, delimiterString)[1]
+	epgDn = strings.Replace(epgDn, "]", "", -1)
+	return epgDn
+}
+
+func (cont *AciController) epgClassChanged(obj apicapi.ApicObject) {
+	dn := obj.GetAttrStr("dn")
+	epgDn := cont.getSubString(dn)
+	cont.indexMutex.Lock()
+	if ok := !cont.contains(cont.cachedEpgDns, epgDn); ok {
+		cont.cachedEpgDns = append(cont.cachedEpgDns, epgDn)
+	}
+	cont.indexMutex.Unlock()
+	sort.Strings(cont.cachedEpgDns)
+}
+
+func (cont *AciController) epgClassDeleted(dn string) {
+	epgDn := cont.getSubString(dn)
+	if cont.contains(cont.cachedEpgDns, epgDn) {
+		cont.indexMutex.Lock()
+		cont.removeSlice(cont.cachedEpgDns, epgDn)
+		cont.indexMutex.Unlock()
+	}
+}
+
+func (cont *AciController) removeSlice(s []string, searchterm string) {
+	i := sort.SearchStrings(s, searchterm)
+	if i < len(s) && s[i] == searchterm {
+		s = append(s[:i], s[i+1:]...)
+	}
+	return
+}
+
+func (cont *AciController) getCachedEpgDns() []string {
+	return cont.cachedEpgDns
```

Comment:
should you be returning a copy here or are all callers expected to use the lock indexMutex

actually the only place I see this used is another member function...couldn't you just acquire the lock and reference cont.cachedEpgDns from there as well?

Reply:
In older code structure, one of the member function asked for a getter to access the cache. Now that I modified the code a bit, I retained it. In case if needed to access the list of epg bound to vmm domain from a diff function, we can use this getter. Sure, I can lock it while referencing this.

Reply:
Well I will take that back. We don't need a getter method since this entity belongs to controller struct. I can access it directly by acquiring a lock from anywhere.